corehq/apps/sms/tests/test_form_session_handler.py

```diff
@@ -151,3 +156,120 @@ def test_incoming_sms_multiple_subevents_for_session(self):
+
+
+@flag_enabled("ONE_PHONE_NUMBER_MULTIPLE_CONTACTS")
+@patch('corehq.apps.smsforms.util.critical_section_for_smsforms_sessions',
+       new=lambda contact_id: contextlib.supress())
+class FormSessionMultipleContactsTestCase(TestCase):
+
+    @classmethod
+    def setUpTestData(cls):
+        cls.domain_name = uuid.uuid4().hex
+
+        cls.backend = SQLTestSMSBackend.objects.create(
+            name='BACKEND',
+            domain=cls.domain_name,
+            is_global=False,
+            hq_api_id=SQLTestSMSBackend.get_api_id()
+        )
+
+        cls.number1 = PhoneNumber.objects.create(
+            domain=cls.domain_name,
+            owner_doc_type='CommCareCase',
+            owner_id='fake-owner-id1',
+            phone_number='01112223333',
+            backend_id='BACKEND',
+            ivr_backend_id=None,
+            verified=True,
+            is_two_way=True,
+            pending_verification=False,
+            contact_last_modified=datetime.utcnow()
+        )
+
+        cls.number2 = PhoneNumber.objects.create(
+            domain=cls.domain_name,
+            owner_doc_type='CommCareCase',
+            owner_id='fake-owner-id2',
+            phone_number='01112223333',
+            backend_id='BACKEND',
+            ivr_backend_id=None,
+            verified=True,
+            is_two_way=True,
+            pending_verification=False,
+            contact_last_modified=datetime.utcnow()
+        )
+
+    def setUp(self):
+        get_channel_for_contact.clear(self.number1.owner_id, self.number1.phone_number)
+        get_channel_for_contact.clear(self.number2.owner_id, self.number2.phone_number)
+
+    @patch('corehq.apps.sms.handlers.form_session.answer_next_question', MagicMock(return_value=None))
+    def test_incoming_sms_linked_form_session__session_contact_matches_incoming(self):
+        session = self._make_session(self.number1)
+        self._claim_channel(session)
+
+        msg = self._get_sms()
+        handled = form_session_handler(self.number1, msg.text, msg)
+        self.assertTrue(handled)
+
+        # message gets attached to the correct session
+        self.assertEqual(msg.xforms_session_couch_id, session.couch_id)
+
+        # sticky session is still 'open'
+        session_info = XFormsSessionSynchronization.get_running_session_info_for_channel(session.get_channel())
+        self.assertEqual(session_info.session_id, session.session_id)
+
+    @patch('corehq.apps.sms.handlers.form_session.answer_next_question', MagicMock(return_value=None))
+    @patch('corehq.apps.sms.handlers.form_session.send_sms_to_verified_number', MagicMock(return_value=None))
+    def test_incoming_sms_linked_form_session__session_contact_different_from_incoming(self):
+        session = self._make_session(self.number2)
+        self._claim_channel(session)
+
+        msg = self._get_sms()
+        # session belongs to `number2` but message comes from `number1`
```

Comment:
Another way to phrase this is that I thought the bug referenced is that the message was associated with the wrong contact, not that it was associated with the right contact but we then applied it incorrectly. This change (and associated test), still seem like the right behavior if we do get into this situation, but I am bit confused how this situation is possible without other bugs happening elsewhere.

Reply:
The main issue is that if the project is using the 'global' gateway with this FF enabled then the FF does not get activated when the [number selection occurs](https://github.com/dimagi/commcare-hq/blob/ea1f477742ada5ba7898c21522dac4e71f70e779/corehq/apps/sms/api.py#L641-L643) because the global gateway has no domain associated with it.

In that instance a session may be active for ContactA (initiated by CC) but the response get's associated with ContactB, however since the logic for finding a session is based purely on `backend_id` and `phone_number` and because the contacts both have the same number, the session is still found even though the incorrect contact has been selected.

I have added a note to the docs about the global gateway issue: https://confluence.dimagi.com/display/saas/One+Phone+Number+-+Multiple+Contacts